pkg/validations/validate.go

```diff
@@ -12,47 +12,174 @@ import (
 }
 
-// Validate runs the assertion from the Rule with the node's value as arguments.
+// validationKwargs represent the optional keyword arguments and their values in a validation annotation.
+type validationKwargs struct {
+	when         *starlark.Callable
+	whenNullSkip bool
+}
+
+// Run takes a root Node, and threadName, and validates each Node in the tree.
+//
+// When a Node's value is invalid, the errors are collected and returned in an AssertCheck.
+// Otherwise, returns empty AssertCheck and nil error.
+func Run(n yamlmeta.Node, threadName string) AssertCheck {
+	if n == nil {
+		return AssertCheck{}
+	}
+
+	validationRunner := newValidationRunner(threadName)
+	err := yamlmeta.Walk(n, validationRunner)
+	if err != nil {
+		return AssertCheck{}
+	}
+
+	return validationRunner.chk
+}
+
+type validationRunner struct {
+	thread *starlark.Thread
+	chk    AssertCheck
+}
+
+func newValidationRunner(threadName string) *validationRunner {
+	return &validationRunner{thread: &starlark.Thread{Name: threadName}, chk: AssertCheck{[]error{}}}
+}
+
+// Visit if `node` has validations in its meta.
+// Runs the validations, any violations from executing the assertions are collected.
+//
+// This visitor stores error(violations) in the validationRunner and returns nil.
+func (a *validationRunner) Visit(node yamlmeta.Node) error {
+	// get rules in node's meta
+	validations := Get(node)
+
+	if validations == nil {
+		return nil
+	}
+	for _, v := range validations {
+		// possible refactor to check validation kwargs prior to validating rules
+		errs := v.Validate(node, a.thread)
+		if errs != nil {
+			a.chk.Violations = append(a.chk.Violations, errs...)
+		}
+	}
+
+	return nil
+}
+
+// Validate runs the assertions in the rules with the node's value as arguments IF
+// the ValidationKwargs conditional options pass.
 //
 // Returns an error if the assertion returns False (not-None), or assert.fail()s.
 // Otherwise, returns nil.
-func (r Rule) Validate(node yamlmeta.Node, thread *starlark.Thread) error {
+func (v NodeValidation) Validate(node yamlmeta.Node, thread *starlark.Thread) []error {
+	key, nodeValue := v.newKeyAndStarlarkValue(node)
+
+	executeRules, err := v.validationKwargs.shouldValidate(nodeValue, thread)
+	if err != nil {
+		return []error{err}
+	}
+	if !executeRules {
+		return nil
+	}
+
+	var failures []error
+	for _, r := range v.rules {
+		result, err := starlark.Call(thread, r.assertion, starlark.Tuple{nodeValue}, []starlark.Tuple{})
+		if err != nil {
+			failures = append(failures, fmt.Errorf("%s (%s) requires %q; %s (by %s)", key, node.GetPosition().AsCompactString(), r.msg, err.Error(), v.position.AsCompactString()))
+		} else if _, ok := result.(starlark.NoneType); !ok {
+			// in order to pass, the assertion must return True or None
+			if !result.Truth() {
+				failures = append(failures, fmt.Errorf("%s (%s) requires %q (by %s)", key, node.GetPosition().AsCompactString(), r.msg, v.position.AsCompactString()))
+			}
+		}
```

Comment:
```suggestion
			_, isNone := result.(starlark.NoneType)
			isTrue := bool(result.Truth())
			if !(isNone || isTrue) {
				failures = append(failures, fmt.Errorf("%s (%s) requires %q (by %s)", key, node.GetPosition().AsCompactString(), r.msg, v.position.AsCompactString()))
			}
		}
```

Even after talking it through, this hunk tripped my internal threshold for cognitive load.

Suggesting we decompose this just slightly to make more explicit what it means for a rule to pass: it's either `None` or `True`-(thy)